Update: report double extra parens in no-extra-parens (fixes #12127)

yeonjuan opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))


<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
Fix #12127 



I tried to solve issue #12127 by adding just `|| hasDoubleExcessParens()` everywhere, but it had a performance issue when checking `ArrayExpression`'s element nodes.

So, I made a new function(`hasExcessParensWithConsiderPrecedence`) which checks extra parens with considering the precedence of a node. This function checks `isParenthesisedTwice` only if the node surrounded by at least one extra paren. 


**Is there anything you'd like reviewers to focus on?**

This pr doesn't change the option's behavior that @mdjermanovic [mentioned](https://github.com/eslint/eslint/issues/12127#issuecomment-522782361).
> There is also a question of options, do they allow just one pair or more.

kaicataldo commented on lib/rules/no-extra-parens.js:
```diff
@@ -169,6 +169,26 @@ module.exports = {
+         * @returns {boolean} True if the node is has an unexpected extra pair of parentheses.
+         * @private
+         */
+        function hasExcessParensWithConsiderPrecedence(node, precedenceLowerLimit) {
```
I wonder if we could rename this to `hasExcessParensWithPrecedence` or ``hasExcessParensWhileConsideringPrecedence` (I prefer the first)?

kaicataldo commented on lib/rules/no-extra-parens.js:
```diff
@@ -169,6 +169,26 @@ module.exports = {
         }
 
+        /**
+         * Determines if a node that is expected to be parenthesised is surrounded by
```
Do you mind expanding on what `considering precedence` means here? Alternatively, a comment below (with code examples) would also be helpful. Specifically, explaining why `isParenthesisedTwice(node)` is being checked here would help clarify the behavior for the next person!

yeonjuan replied:
@kaicataldo 
yes! it makes sense. thanks! I changed it.

mdjermanovic commented on lib/rules/no-extra-parens.js:
```diff
@@ -169,6 +169,28 @@ module.exports = {
+         * Determines if a node that is expected to be parenthesised is surrounded by
+         * (potentially) invalid extra parentheses with considering precedence level of the node.
+         * If the preference level of the node is higher or equal to precedence lower limit, it checks
+         * whether the node is surrounded by parentheses twice or not.
```
This sounds more like `&&` than `||` :-)

yeonjuan replied:
@mdjermanovic 
yes :) you are right. I will change the comments. Thanks 👍

yeonjuan commented on lib/rules/no-extra-parens.js:
```diff
@@ -429,7 +451,9 @@ module.exports = {
-                .filter(arg => hasExcessParens(arg) && precedence(arg) >= PRECEDENCE_OF_ASSIGNMENT_EXPR)
+                .filter(arg => hasExcessParens(arg))
+                .filter(arg => precedence(arg) >= PRECEDENCE_OF_ASSIGNMENT_EXPR ||
+                    hasDoubleExcessParens(arg))
```
Sure. Thanks!

yeonjuan commented on lib/rules/no-extra-parens.js:
```diff
@@ -448,11 +472,22 @@ module.exports = {
+            if (!shouldSkipLeft && hasExcessParens(node.left)) {
+                if (
+                    (leftPrecedence > prec || (leftPrecedence === prec && !isExponentiation)) ||
+                    hasDoubleExcessParens(node.left)
```
@mdjermanovic 
Yes. that makes sense.
That would be more clear. Besides, that's reasonable because the `hasDoubleExcessParens` has the duplicate logic(checking `ruleApplies`) with `hasExcessParens`

mdjermanovic commented:
> This pr doesn't change the option's behavior that @mdjermanovic [mentioned](https://github.com/eslint/eslint/issues/12127#issuecomment-522782361).
>> There is also a question of options, do they allow just one pair or more.

By the docs, it sounds like any number of extra parens should be allowed when something is excepted by an option.

> * "conditionalAssign": false allows extra parentheses around assignments in conditional test expressions
> * "returnAssign": false allows extra parentheses around assignments in return statements
> ...

There are also cases where double parens are indeed required. For example, `conditionalAssign` was added to avoid conflicts with `no-cond-assign` (#3317), and that rule allows assignments in conditional expressions' test conditions only if they're  parenthesized twice:

```js
/*eslint no-cond-assign: "error"*/

var a = ((b = c)) ? foo : bar; // ok

var a = (b = c) ? foo : bar; // not ok
```

So it's probably best to leave that part as is.

mdjermanovic commented:
Ooops, `ConditionalExpression(node)` doesn't actually check `conditionalAssign` option, while it probably should.

But it does check `returnAssign` option, it's might be relevant to read #6095.

yeonjuan commented:
@mdjermanovic 

> Ooops, ConditionalExpression(node) doesn't actually check conditionalAssign option, while it probably should.

Oh. yes. I got it. The problem has not been discovered because the bug(#12127) allows extra twice parens.
``` js
// Error in the current PR
 {
   code: "var a = ((b = c)) ? foo : bar;",
   options: ["all", { conditionalAssign: false }]
 }
```
I can fix it, but maybe there are some other issues. so can I add more test cases for `option`?

mdjermanovic commented:
> Oh. Yes. I got it. The problem has not been discovered because the bug(#12127) allows extra twice parens in those cases.
> 
> ```js
> // Error in the current PR
>  {
>    code: "var a = ((b = c)) ? foo : bar;",
>    options: ["all", { conditionalAssign: false }]
>  }
> ```

Yes, it seems that this wasn't noticed because of #12127 bug. 

Now when the bug is going to be fixed, we should ensure that this code stays `valid` for `conditionalAssign: true`.

> I can fix it, but maybe there are some other undiscovered issues. So can I add more test cases for `option`?

That would be great, thanks!

yeonjuan commented:
@mdjermanovic  
I fixed it by adding `isCondAssignException()`.
By the way, what do you think about this case? - [demo - enforceForArrowConditionals: false](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWV4dHJhLXBhcmVuczogW1wiZXJyb3JcIiwgXCJhbGxcIiwgeyBcImVuZm9yY2VGb3JBcnJvd0NvbmRpdGlvbmFsc1wiOiBmYWxzZSB9XSAqL1xuXG52YXIgYSA9ICgpID0+ICgxID8gMiA6IDMpO1xudmFyIGEgPSAoKSA9PiAoKDEgPyAyIDogMykpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19) 
```js
/* eslint no-extra-parens: ["error", "all", { "enforceForArrowConditionals": false }] */

var a = () => (1 ? 2 : 3);
var a = () => ((1 ? 2 : 3));  // Unnecessary parentheses arround expression.
``` 
The option(`enforceDorArrowConditionals`) doesn't allow extra twice parens in the actual version.
I think it can be treated in separate PR because it occurs in the actual version and less related to this PR, but I want to know whether it is a bug or not.

mdjermanovic commented:
> The option(`enforceDorArrowConditionals: false`) doesn't allow extra twice parens in the actual version.
> I think it can be treated in separate PR because it occurs in the actual version and less related to this PR, but I want to know whether it is a bug or not.

I agree, fixing that behavior can be a separate PR (if it's a bug at all).

It looks like the options are inconsistent about this. Some allow only 1 pair of extra parens while the others allow 1, 2, or more. The documentation isn't very specific about this, although it sounds more like that any number of extra parens should be allowed.

It's hard to tell what makes more sense. Probably only 1, except for the special cases such as conditional expressions where the presence of 2 pairs has a meaning for another rule. On the other hand, an option name such as `ignoreJSX` might imply that the rule should really completely ignore JSX and allow any number of extra parens.

yeonjuan commented on lib/rules/no-extra-parens.js:
```diff
@@ -702,25 +731,24 @@ module.exports = {
 
             ConditionalExpression(node) {
-                if (isReturnAssignException(node)) {
+                if (isReturnAssignException(node) || isCondAssignException(node)) {
```
@mdjermanovic  Ooops, Thanks. You are definitely right.
I changed it and added test-cases for preventing regression.

mdjermanovic replied:
Thanks, that was really fast!

mdjermanovic commented:
`function checkUnaryUpdate(node)` might also need to be changed? E.g., `-((a + b))`

It could be nice to change `SequenceExpression(node)` as well, although it's untestable at the moment because single parens are always unnecessary with the actual syntax.

yeonjuan commented:
@mdjermanovic Yes!. That should be.
And `VariableDeclarator` needs to be changed also. - ex) `var foo = ((bar, baz));`
I'll check all the cases one more and commit. thanks!

yeonjuan commented:
@mdjermanovic It seems to cover all.
To note,
1. updated for checking double parens
 - in `checkUnaryUpdate`
 - in `SequenceExpression`
 - in `VariableDeclarator`
2. updated for refactoring
 -  in `checkSpreadOperator`
 -  in `checkClass`

mdjermanovic commented:
Actually, let's do `MemberExpression(node)` in a separate PR. 

This PR has already a lot of changes, and it seems that the logic in `MemberExpression(node)` needs to be modified (#12740).

yeonjuan commented:
@mdjermanovic 
Wow. it has really many cases for checking. Thanks, I'll check all once  again including what you mentioned

yeonjuan commented on lib/rules/no-extra-parens.js:
```diff
@@ -379,8 +401,7 @@ module.exports = {
-            if (hasExcessParens(node.argument) && precedence(node.argument) >= precedence(node)) {
+            if (hasExcessParensWithPrecedence(node.argument, precedence(node))) {
                 report(node.argument);
             }
```
@mdjermanovic 
> I think this if is actually an unnecessary code.
Yes! I checked it. it's right. Thanks :)

mdjermanovic commented on lib/rules/no-extra-parens.js:
```diff
@@ -470,11 +499,7 @@ module.exports = {
-                : hasDoubleExcessParens(node.superClass);
-
-            if (hasExtraParens) {
+            if (hasExcessParensWithPrecedence(node.superClass, PRECEDENCE_OF_UPDATE_EXPR)) {
```
This changes checking precedence from `>` to `>=`, so the rule would now remove parens and produce parsing error in e.g.:

```js
class A extends (++x) {}
```

mdjermanovic commented:
One edge case: `for (a of ((b, c)));`

mdjermanovic commented:
Just changed the title to `Update:` because this change produces more warnings.

kaicataldo commented:
Thanks for contributing to ESLint!
